File: theo-algo-komplexitaet/code/sat.java

// Intended to be run using Java > 23 (Tested with Java 23 and 24)
// May require --enable-preview to do so.

interface Expr {

    /**
     * An Optional True or Optional False if this expression 
     * definitively evaluates to the respective truth value 
     * with the given binding or an empty Optional otherwise.
     * 
     * Returning a truth value does not necessarily
     * require all variables to be bound to a definite
     * value. For example, Optional.empty will be returned,
     * if the truth value cannot be determined with the given
     * binding. E. g., if this expression represents a
     * variable for which the binding has no value,
     * Optional.empty is returned.
     * 
     * An expression such as "A ⋀ B" would return true
     * if A and B are both bound to "true"  and false 
     * if at least one of them is bound
     * to "false", and Optional.empty otherwise.
     */
    Optional<Boolean> isSolution(Map<Var, Boolean> binding);
}

class And implements Expr {

    private final Expr[] exprs;

    And(Expr... exprs) {
        this.exprs = exprs;
    }

    public Optional<Boolean> isSolution(Map<Var, Boolean> binding) {
        Optional<Boolean> r = Optional.of(true);
        for (var expr : this.exprs) {
            final var e = expr.isSolution(binding);
            if (!e.isPresent())
                r = Optional.empty();
            else if (!e.get())
                return e;
        }
        return r;
    }

    public String toString() {
        return Arrays.stream(exprs)
                .map(Expr::toString)
                .collect(Collectors.joining(" ⋀ "));
    }
}

class Or implements Expr {

    private final Expr[] exprs;

    Or(Expr... exprs) {
        this.exprs = exprs;
    }

    public Optional<Boolean> isSolution(Map<Var, Boolean> binding) {
        var r = Optional.of(false);
        for (var expr : this.exprs) {
            final var e = expr.isSolution(binding);
            if (!e.isPresent())
                r = Optional.empty();
            else if (e.get())
                return e;
        }
        return r;
    }

    public String toString() {
        return Arrays.stream(exprs)
                .map(Expr::toString)
                .collect(Collectors.joining(" ⋁ "));
    }
}

class Not implements Expr {

    private final Expr expr;

    Not(Expr expr) {
        this.expr = expr;
    }

    public Optional<Boolean> isSolution(Map<Var, Boolean> binding) {
        final var r = expr.isSolution(binding).map(b -> !b);
        return r;
    }

    public String toString() {
        return "¬" + expr;
    }
}

class Var implements Expr {

    private final String name;

    Var(String name) {
        this.name = name;
    }

    public Optional<Boolean> isSolution(Map<Var, Boolean> binding) {
        final var r = Optional.ofNullable(binding.get(this));
        return r;
    }

    public String toString() {
        return name;
    }

}

void main() {
    final Var A = new Var("a");
    final Var B = new Var("b");
    final Var C = new Var("c");
    final Var D = new Var("d");
    Stack<Var> vars = new Stack<>();
    vars.addAll(Arrays.asList(new Var[] { A, B, C, D }));
    Expr expr = new And(
            new Or(A, B),
            new Or(new Not(A), B),
            new Or(new Not(A), new Not(C)),
            new Or(C, D),
            new Or(new Not(C), new Not(D)));
    println("Finding solutions for: " + expr.toString());

    Map<Var, Boolean> solution = new HashMap<>();
    solve(expr, vars, solution);
}

void solve(Expr expr, Stack<Var> vars, Map<Var, Boolean> solution) {
    final var var = vars.pop();
    for (var v : new Boolean[] { true, false }) {
        solution.put(var, v);

        final var e = expr.isSolution(solution);
        // println("Current solution: " + solution+ " => "+ e);
        if (!e.isPresent()) {
            solve(expr, vars, solution);
        } else if (e.get()) {
            println("Solution found: " + solution);
        }
    }
    vars.push(var);
    solution.remove(var);
}

File: theo-algo-komplexitaet/code/sat.py
from abc import abstractmethod

class Expr:

    @abstractmethod
    def is_solution(
            self, 
            binding: dict["Var", bool]) -> bool | None:
        """True or False if this expression definitively
        evaluates to the respective truth value with the
        given binding or None otherwise.

        Returning a truth value does not necessarily
        require all variables to be bound to a definite
        value.

        For example, None will be returned, if the
        truth value cannot be determined with the given
        binding. E. g., if this expression represents a
        variable for which the binding has no value, None
        is returned.

        An expression such as "A ⋀ B" would return True
        if A and B are both True in the
        binding and False if at least one of them is bound
        to False, and None otherwise.
        """
        raise NotImplementedError


class And(Expr):
    def __init__(self, *exprs: Expr):
        self.exprs = exprs

    def is_solution(self, binding):
        r = True
        for expr in self.exprs:
            e = expr.is_solution(binding)
            if e is None:
                r = None
            elif not e:
                return False
        return r

    def __str__(self):
        return " ⋀ ".join(str(expr) for expr in self.exprs)

    def __repr__(self):
        exprs = ", ".join(str(expr) for expr in self.exprs)
        return "And(" + exprs + ")"


class Or(Expr):
    def __init__(self, *exprs: Expr):
        self.exprs = exprs

    def is_solution(self, binding):
        r = False
        for expr in self.exprs:
            e = expr.is_solution(binding)
            if e is None:
                r = None
            elif e:
                return True
        return r

    def __str__(self):
        return " ⋁ ".join(str(expr) for expr in self.exprs)

    def __repr__(self):
        exprs = ", ".join(repr(expr) for expr in self.exprs)
        return "Or(" + exprs + ")"


class Not(Expr):
    def __init__(self, expr: Expr):
        self.expr = expr

    def is_solution(self, binding):
        e = self.expr.is_solution(binding)
        if e is None:
            return None
        else:
            return not e

    def __str__(self):
        return f"¬{self.expr}"

    def __repr__(self):
        return "Not(" + repr(self.expr) + ")"


class Var(Expr):
    def __init__(self, name: str):
        self.name = name

    def is_solution(self, binding):
        """True or False if bound.
        None if unbound (default).
        """
        if self not in binding:
            return None
        else:
            return binding[self]

    def __str__(self):
        return self.name

    def __repr__(self):
        return 'Var("' + self.name + '")'


A = Var("a")
B = Var("b")
C = Var("c")
D = Var("d")
vars = [A, B, C, D]
""" The variables are now indexed to enable iterating over 
    them in the solve function. """

expr = And(
    Or(A, B),
    Or(Not(A), B),
    Or(Not(A), Not(C)),
    Or(C, D),
    Or(Not(C), Not(D)),
)
print("Finding solutions for: " + expr.__str__())

solution: dict[Var, bool] = {}
""" Stores the current solution by mapping the name of a 
    variable to its current truth value (True or False)."""


def solve(expr, vars):
    var = vars[0]
    for v in [True, False]: # try both values for the variable
        solution[var] = v # bind the variable to the value
        e = expr.is_solution(solution) # check if the expression
                                       # is a solution
        # print("State: " + str(solution) + " => " + str(e))
        if e is None: # not inconsistent -> continue
            solve(expr, vars[1:]) # try the next variable
        elif e:
            print("Solution found: " + str(solution))

    del solution[var] # enable backtracking


solve(expr, vars)
#solve(And(Or(A, B), C, Not(B)), vars)
